### ISSUE TITLE
How to determine if decode failure is due to needing more bytes

### ISSUE THREAD

**Issue by ljbade** [gabhijit/hampi#103]
The asn1c library provides a way to detect when an array of bytes is too short foe decoding the data. This might happen if you are reading in multiple concatenated messages and you don't know the number of bytes  of each message.

Thus in asn1c you can just keeping reading more bytes until you have the complete message and decoding succeeds.

I don't see a way to do this with the current decoder as the `PerCodecError` type doesn't seem to have a way to indicate that more bytes are required.

**Comment by gabhijit** [comment]
If I understand your question correctly, you are saying that `PerCodecError` does not give an indication of 'what' caused error, Other than the error message, right? May be something like `cause` can be added to `PerCodecError`.  There might even be places where it simply says `PerCodecError` and nothing else! 

Is this correct? Or have I missed something?

**Comment by ljbade** [comment]
Yes at a minimum I would like a way to determine if the error from a decode function is due to something in `PerCodecData` finding there is not enough bytes to decode a field versus something else like a corrupt message.

A cause field or an enum would be good, as then you can check the type of error, similar to the Rust standard library errors.

**Comment by gabhijit** [comment]
@ljbade: Sure will add that

**Comment by gabhijit** [comment]
@ljbade : Here is the current implementation - Added a field `cause` to the `PerCodecError` struct. This is an enum that looks like the following 

```rust
#[derive(Debug)]
pub enum ErrorCause {
    /// Remaining Buffer is too short to decode
    BufferTooShort,

    /// Alignment Error during decode
    InvalidAlignment,

    /// Encoding of value is not currently supported
    EncodeNotSupported,

    /// Decoding of value is not currently supported
    DecodeNotSupported,

    /// Generic Error during Encode or Decode
    Generic,
}
```
This is implemented in #104. Which is now merged in master. You might want to test with master branch and see if this makes sense for you. If this looks good, you may want to close the issue.

Once this and #101 are closed, I will make a 0.6.0 release. Since this is a change in the `PerCodecError` API.

**Comment by gabhijit** [comment]
@ljbade : I am not sure whether you've got a chance to look at this. I would wait for a few more days and mark this issue as closed.

**Comment by gabhijit** [comment]
This is available now in crate version `0.6.0`
